tools/tpm2_tool.c

```diff
@@ -61,18 +62,101 @@ static ESYS_CONTEXT* ctx_init(TSS2_TCTI_CONTEXT *tcti_ctx) {
     return esys_ctx;
 }
 
+/*
+ * Build a list of the TPM2 tools linked into this executable
+ */
+static const tpm2_tool ** tools;
+static unsigned tool_count;
+
+void tpm2_tool_register(const tpm2_tool * tool)
+{
+    if (!tools)
+        tools = calloc(1024, sizeof(*tools));
+
+    //printf("%s: register %s\n", __func__, tool->name);
+    if (tools)
+        tools[tool_count++] = tool;
+}
+
+static const char * tpm2_tool_name(const char * arg)
+{
+    const char * name = rindex(arg, '/');
+    if (name)
+        name++; // skip the '/'
+    else
+        name = arg; // use the full executable name as is
+
+    if (strncmp(name, "tpm2_", 5) == 0)
+        name += 5;
+
+    return name;
+}
+
+static const tpm2_tool * tpm2_tool_lookup(int *argc, char *** argv)
+{
+    // no tools linked in?
+    if (!tools)
+        return NULL;
+
+    // only one tool linked in? use it no matter the program name.
+    if (tool_count == 1)
+        return *tools;
+
+    // find the executable name in the path
+    // and skip "tpm2_" prefix if it is present
+    const char * name = tpm2_tool_name((*argv)[0]);
+
+    // if this was invoked as 'tpm2', then try again with the second argument
+    if (strcmp(name, "tpm2") == 0)
+    {
+        if (--(*argc) == 0)
+            return NULL;
+        (*argv)++;
+        name = tpm2_tool_name((*argv)[0]);
+    }
+
+
+    // search the tools array for a matching name
+    for(unsigned i = 0 ; i < tool_count ; i++)
+    {
+        const tpm2_tool * const tool = tools[i];
+        if (!tool || !tool->name)
+            continue;
+        if (strcmp(name, tool->name) == 0)
+            return tool;
+    }
+
+    // not found? should print a table of the tools
+    return NULL;
+}
+
+
 /*
  * This program is a template for TPM2 tools that use the SAPI. It does
  * nothing more than parsing command line options that allow the caller to
  * specify which TCTI to use for the test.
  */
-int main(int argc, char *argv[]) {
```

Comment:
Don't change stuff like this, here and throughout. We do lists of pointers with the array brackets.

Reply:
This change was necessary because `tpm2_tool_lookup()` modified the command line based on if it was invoked as `tpm2` or as `tpm2_foo`.  With `char * argv[]` it is not possible to do the `argv++` to skip the 0th argument.